File: api/torrent_info.go
package api

import (
	"encoding/json"
	"net/http"

	real_debrid "github.com/sushydev/real_debrid_go"
)

type TorrentFile struct {
	ID       int    `json:"id"`
	Path     string `json:"path"`
	Bytes    int    `json:"bytes"`
	Selected int    `json:"selected"`
}

type TorrentInfo struct {
	ID               string        `json:"id"`
	Filename         string        `json:"filename"`
	OriginalFilename string        `json:"original_filename"`
	Hash             string        `json:"hash"`
	Bytes            int           `json:"bytes"`
	OriginalBytes    int           `json:"original_bytes"`
	Host             string        `json:"host"`
	Split            int           `json:"split"`
	Progress         float64       `json:"progress"`
	Status           string        `json:"status"`
	Added            string        `json:"added"`
	Files            []TorrentFile `json:"files"`
	Links            []string      `json:"links"`
	Ended            string        `json:"ended,omitempty"`
	Speed            int           `json:"speed,omitempty"`
	Seeders          int           `json:"seeders,omitempty"`
}

func GetTorrentInfo(client *real_debrid.Client, id string) (*TorrentInfo, error) {
	url := client.GetUrl("/torrents/info/" + id)

	req, err := http.NewRequest("GET", url.String(), nil)
	if err != nil {
		return nil, err
	}

	response, err := client.Do(req)
	if err != nil {
		return nil, err
	}

	defer response.Body.Close()

	err = client.HandleResponseCode(response, 200)
	if err != nil {
		return nil, err
	}

	torrentInfo := &TorrentInfo{}

	err = json.NewDecoder(response.Body).Decode(torrentInfo)
	if err != nil {
		return nil, err
	}

	return torrentInfo, nil
}

File: lib/real_debrid/api/torrent_info.ex
defmodule RealDebrid.Api.TorrentInfo do
  @moduledoc """
  Get detailed torrent information.
  """

  alias RealDebrid.Client

  @type torrent_file :: %{
          id: integer(),
          path: String.t(),
          bytes: integer(),
          selected: integer()
        }

  @type torrent_info :: %{
          id: String.t(),
          filename: String.t(),
          original_filename: String.t(),
          hash: String.t(),
          bytes: integer(),
          original_bytes: integer(),
          host: String.t(),
          split: integer(),
          progress: integer(),
          status: String.t(),
          added: String.t(),
          files: [torrent_file()],
          links: [String.t()],
          ended: String.t() | nil,
          speed: integer() | nil,
          seeders: integer() | nil
        }

  @doc """
  Get detailed information about a torrent.

  ## Parameters

    - `client` - The Real-Debrid client
    - `id` - The torrent ID

  ## Returns

    - `{:ok, torrent_info}` on success
    - `{:error, reason}` on failure
  """
  @spec get(Client.t(), String.t()) :: {:ok, torrent_info()} | {:error, term()}
  def get(%Client{} = client, id) do
    case Client.get(client, "/torrents/info/#{id}") do
      {:ok, body, _headers} ->
        {:ok, parse_torrent_info(body)}

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp parse_torrent_info(body) do
    %{
      id: body["id"],
      filename: body["filename"],
      original_filename: body["original_filename"],
      hash: body["hash"],
      bytes: body["bytes"],
      original_bytes: body["original_bytes"],
      host: body["host"],
      split: body["split"],
      progress: body["progress"],
      status: body["status"],
      added: body["added"],
      files: parse_files(body["files"] || []),
      links: body["links"] || [],
      ended: body["ended"],
      speed: body["speed"],
      seeders: body["seeders"]
    }
  end

  defp parse_files(files) do
    Enum.map(files, fn file ->
      %{
        id: file["id"],
        path: file["path"],
        bytes: file["bytes"],
        selected: file["selected"]
      }
    end)
  end
end
